warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java

```diff
@@ -0,0 +1,387 @@
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.ASTReferenceExpression;
+import org.apache.commons.jexl2.parser.ASTReturnStatement;
+import org.apache.commons.jexl2.parser.ASTSizeFunction;
+import org.apache.commons.jexl2.parser.ASTSizeMethod;
+import org.apache.commons.jexl2.parser.ASTStringLiteral;
+import org.apache.commons.jexl2.parser.ASTTernaryNode;
+import org.apache.commons.jexl2.parser.ASTTrueNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTVar;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.SimpleNode;
+
+/**
+ * Validates that every ASTAndNode and every ASTOrNode has more than one child
+ */
+public class JunctionValidatingVisitor extends BaseVisitor {
+    
+    private boolean isValid = true;
+    
+    public static boolean validate(JexlNode script) {
+        JunctionValidatingVisitor visitor = new JunctionValidatingVisitor();
+        script.jjtAccept(visitor, null);
+        return visitor.getIsValid();
+    }
+    
+    public boolean getIsValid() {
+        return this.isValid;
+    }
+    
+    /*
+     * Pass through
+     */
+    
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTReference node, Object data) {
+        if (!isValid) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
+        if (!isValid) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(SimpleNode node, Object data) {
+        if (!isValid) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    // check the argument nodes of a function
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        if (!isValid) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    // descend into negated branches of the query
+    @Override
+    public Object visit(ASTNotNode node, Object data) {
+        if (!isValid) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    /*
+     * Validation methods
+     */
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        if (!isJunctionValid(node)) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        if (!isJunctionValid(node)) {
```

Comment:
Just a thought here, let me know what you think.

If I read this method, it's not clear that the isValid flag is getting set unless I jump into `isJunctionValid()`. The `isJunctionValid()` method is doing double-duty. It both checks if the node is valid and has a side-effect. If you broke those two apart, this method could be:

```
public Object visit(ASTAndNode node, Object data) {
  if (!isValid)
    return data;
  else if (!isJunctionValid(node)) {
    isValid = false;
    return data;
  }
  // recurse down tree
}
```

Then `isJunctionValid()` only has a single job of checking a node's validity. It's cool if you wanna skip this but wanted to mention it.

Reply:
Unlike #1299 the validate methods are the same for and/or nodes, going to extract this code to a common method. Good call!